woocommerce/woocommerce
# Fix error notice dismissal on checkout

### Changes proposed in this Pull Request:

Fixes #52813

### Description of changes

There are 2 sets of errors present on the `Cart` state.

1. At the root there is `state.errors`
2. There is also `state.cartData.errors`

`2` Is errors that come directly with the cart response which we set via `receiveCart`.

`1` Is other response errors, we set these for various conditions. One of those conditions is when we update the customer via `updateCustomerData`. Updating the customer may result in errors, but before we were only clearing `2` and not `1` if those errors went away.

Furthermore we were trying to dismiss notices from raw API error responses, but when we create the notices some manipulation is done to create a context that is not the same as the raw API error response.

What I've tried to do here is to ensure that:

1. When clearing notices for the errors we are searching for the notice correctly based on the same way the notice was created.

2. We collect errors from both `1` and `2` to ensure that all errors can be cleared if necessary.

In future we should consider refactoring error handling. It may be possible to maintain all error state in one place and also handle the clearing of it from the same place even if errors are received in different ways.

### How to test the changes in this Pull Request:

You can follow the reproduction steps outlined in #52813 (Please watch the video because it shows the exact way you can end up in that state).

Steps to reproduce
* [Copy the example](https://developer.woocommerce.com/docs/cart-and-checkout-additional-checkout-fields/#7-a-full-example) here somewhere or in a snippet plugin.
* Go to Checkout, fill out all fields, when you get to gov id, fill in one but don't confirm to fill out the second
* take focus off the form field you just filled
* Wait to see the error notice appear at top about the gov id field
* Now fix the error by filling in the other gov id field to match the first
* Notice that the error notice **goes away** once you fill in the fields correctly.



### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Fix a bug where server validation errors were not dismissed after validation passes.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
